Add lbann cmake flags

bvanessen opened the pull request:
Update the LBANN CMake flags to be in-line with current CMake options.  Removed `type='build'` flag from dependencies to ensure that dependent packages are placed in the spack environment.

bvanessen committed:
Adding support for the CMake flags in LBANN that are missing.

bvanessen committed:
Added new flag to OpenCV dependency and removed negative variants
since OpenCV no longer turns on everything by default.  Removed CMake
flags in LBANN that have been deprecated.

bvanessen committed:
Removed type='build' flags from dependencies so that they get linked
into a environment's view.

bvanessen committed:
Removed type='build' flags from dependencies so that they get linked
into a environment's view.  Fixed DiHydrogen variant to enable
DistConv feature, renamed to +distconv from +legacy.  Added conflicts
line to indicated that DistConv and ROCm don't work with +half
support.

bvanessen committed:
Fixed Flake8 and cleaned up ordering of variants.

bvanessen committed:
Flake8

bvanessen commented:
@adamjstewart Would it be possible to take a look at this PR today?

adamjstewart commented:
> Removed type='build' flag from dependencies to ensure that dependent packages are placed in the spack environment.

Could you explain this in more detail? Most packages that depend on cmake/ninja only need them at build time, they can be uninstalled after the package is built.

bvanessen commented:
> > Removed type='build' flag from dependencies to ensure that dependent packages are placed in the spack environment.
> 
> Could you explain this in more detail? Most packages that depend on cmake/ninja only need them at build time, they can be uninstalled after the package is built.

I am trying to get a `spack install --only dependencies` command to create an environment that allows you to have everything that you need in it.  Turns out that packages that are only required for `build` are not included in the environment's view.  As a result when you activate the environment you will not see the versions of `cmake` or `ninja` that were required to build the package.  I removed the flags to make sure that they show up in the environment.  The alternative is to use the `spack build-env` to setup the right environment, but since that requires the exact set of variants to get the correct environment it is more cumbersome that just activating a spack environment.

adamjstewart commented:
I don't think this is the right way to do things. If `spack install --only dependencies` is skipping build dependencies, we should fix that. One of the problems with things like `cmake` as a link dependency is that it forces every package in the DAG to be built with the latest and greatest version of CMake. If it's a build dependency, the version of `cmake` doesn't get factored into the hash.

bvanessen commented:
> I don't think this is the right way to do things. If `spack install --only dependencies` is skipping build dependencies, we should fix that. One of the problems with things like `cmake` as a link dependency is that it forces every package in the DAG to be built with the latest and greatest version of CMake. If it's a build dependency, the version of `cmake` doesn't get factored into the hash.

Hmmm, interesting.  I wasn't aware of that.  @becker33 @tgamblin @scheibelp thoughts on how we should handle this?

bvanessen committed:
Backed out changes to not mark and cmake and ninja as build
dependencies, which was introduced to make sure that they appear in
a spack environment.

bvanessen commented:
> > I don't think this is the right way to do things. If `spack install --only dependencies` is skipping build dependencies, we should fix that. One of the problems with things like `cmake` as a link dependency is that it forces every package in the DAG to be built with the latest and greatest version of CMake. If it's a build dependency, the version of `cmake` doesn't get factored into the hash.
> 
> Hmmm, interesting. I wasn't aware of that. @becker33 @tgamblin @scheibelp thoughts on how we should handle this?

@adamjstewart I have removed the changes to strip out the `type=build` flags from `cmake` and `ninja` to separate concerns in this PR.  The PR now focuses on updating the `cmake` flags that we need across the software stack.

adamjstewart commented on var/spack/repos/builtin/packages/dihydrogen/package.py:
```diff
@@ -105,14 +106,17 @@ class Dihydrogen(CMakePackage, CudaPackage):
 
-    depends_on('py-breathe', type='build', when='+docs')
-    depends_on('doxygen', type='build', when='+docs')
+    depends_on('py-breathe', when='+docs')
+    depends_on('doxygen', when='+docs')
```
I assume these are also only needed at build-time

adamjstewart commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -33,31 +33,38 @@ class Lbann(CMakePackage, CudaPackage):
+    variant('opencv', default=True,
+            description='Builds with support for image processing with OpenCV')
+    variant('vtune', default=False, description='Builds with support for Intel VTune')
+    variant('unit_testing', default=False, description='Builds with support for unit testing')
```
Just to make sure, this variant changes what gets installed, right? It doesn't just build unit tests that can be used at build-time?

adamjstewart replied:
Still looking for an answer to this.

bvanessen replied:
It turns on a dependence for catch and builds the catch test, so I have removed the variant and tied into the `--test` flag.

adamjstewart commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -150,12 +156,12 @@ class Lbann(CMakePackage, CudaPackage):
-    depends_on('py-breathe', type='build', when='+docs')
-    depends_on('doxygen', type='build', when='+docs')
-    depends_on('py-m2r', type='build', when='+docs')
+    depends_on('py-breathe', when='+docs')
+    depends_on('doxygen', when='+docs')
+    depends_on('py-m2r', when='+docs')
```
Keep `type='build'`

bvanessen committed:
Backed out changes to not mark doc related packages as build
dependencies, which was introduced to make sure that they appear
in a spack environment.

bvanessen committed:
Fixed how recipe communicates the intent to build and run tests to the
package CMake.
